src/main/java/teammates/ui/template/AdminActivityLogTableRow.java

```diff
@@ -0,0 +1,211 @@
+    private static final String ACTION_UNSCCUSSFUL_HIGHLIGHTER_BACK = "</strong><span>";
+    
+    private static final String KEYWORDS_HIGHLIGHTER_FRONT = "<mark>";
+    private static final String KEYWORDS_HIGHLIGHTER_BACK = "</mark>";
+    
+    private ActivityLogEntry activityLog;
+    
+    private String[] keyStringsToHighlight;
+    
+    public AdminActivityLogTableRow(ActivityLogEntry entry) {
+        activityLog = entry;
+    }
+    
+    public void setKeyStringsToHighlight(String[] strs) {
+        keyStringsToHighlight = strs;
+    }
+    
+    // --------------- Additional generated fields ---------------
+    
+    public String getUserHomeLink() {
+        switch (activityLog.getRoleWithoutMasquerade()) {
+        case Const.ActivityLog.ROLE_STUDENT:
+            return Url.addParamToUrl(Const.ActionURIs.STUDENT_HOME_PAGE,
+                    Const.ParamsNames.USER_ID, activityLog.getGoogleId());
+        case Const.ActivityLog.ROLE_INSTRUCTOR:
+            return Url.addParamToUrl(Const.ActionURIs.INSTRUCTOR_HOME_PAGE,
+                    Const.ParamsNames.USER_ID, activityLog.getGoogleId());
+        default:
+            return null;
+        }
+    }
+    
+    public boolean getHasUserHomeLink() {
+        return activityLog.getRoleWithoutMasquerade().contains(Const.ActivityLog.ROLE_STUDENT)
+                || activityLog.getRoleWithoutMasquerade().contains(Const.ActivityLog.ROLE_INSTRUCTOR);
+    }
+    
+    public String getUserIdentity() {
+        String googleId = activityLog.getGoogleId();
+        if (!googleId.contentEquals(Const.ActivityLog.AUTH_NOT_LOGIN)
+                && !googleId.contentEquals(Const.ActivityLog.UNKNOWN)) {
+            return googleId;
+        }
+        
+        String email = activityLog.getEmail();
+        if (email != null && !email.contentEquals(Const.ActivityLog.UNKNOWN)) {
+            return email;
+        }
+        
+        String name = activityLog.getName();
+        if (name != null && !name.contentEquals(Const.ActivityLog.UNKNOWN)) {
+            return name;
+        }
+        return "";
+    }
+    
+    // --------------- 'is' fields to determine css class ---------------
+    
+    public boolean getIsUserAdmin() {
+        return activityLog.getRole().contains(Const.ActivityLog.ROLE_ADMIN);
+    }
+    
+    public boolean getIsUserInstructor() {
+        return activityLog.getRole().contains(Const.ActivityLog.ROLE_INSTRUCTOR);
+    }
+    
+    public boolean getIsUserStudent() {
+        return activityLog.getRole().contains(Const.ActivityLog.ROLE_STUDENT);
+    }
+    
+    public boolean getIsUserAuto() {
+        return activityLog.getRole().contains(Const.ActivityLog.ROLE_AUTO);
+    }
+    
+    public boolean getIsUserUnregistered() {
+        return activityLog.getRole().contains(Const.ActivityLog.ROLE_UNREGISTERED);
+    }
+    
+    public boolean getIsTimeTakenWarning() {
```

Comment:
Following the previous comment, you may want to rename this `getIsActionTimeModeratelySlow` or similar.